src/Nri/Ui/Modal/V11.elm

```diff
@@ -116,20 +132,19 @@ view model =
                 Modal.view
                     { title = "First kind of modal"
                     , wrapMsg = ModalMsg
-                    , focusManager =
-                        Modal.MultipleFocusableElements
-                            (\{ firstFocusableElement, lastFocusableElement, closeButton } ->
-                                { content =
-                                    [ closeButton firstFocusableElement
-                                    , text "Modal Content"
-                                    ]
-                                , footer =
-                                    [ button
-                                        (Events.onClick CloseModal :: lastFocusableElement)
-                                        [ text "Close" ]
-                                    ]
-                                }
+                    , content =
```

Comment:
I like the change! I liked how the previous API was a bit more strict but this one is definitely more ergonomic!

Reply:
💭 I wonder if something like the following would be feasable as a compromise between both:

```elm
Modal.view
  {...
  -- no need to use firstFocusable and friends in the view. just set the ids.
  , content : []
  , footer = []
  }
  [ -- set the focus manager as an attribute, separated from the actual content
    Modal.singleFocusableElement "element-id"
  , -- alternative for multiple focusable elements
    Modal.multipleFocusableElements
      { firstFocusableElement = "first-focusable-element"
      , lastFocusableElement = "last-focusable-element"
      }
  ]
```

Of course, the user could still use it in a way that the supplied ids point nowhere in the view, but I like that it's a bit more explicit about how focus management works, and there's no room for the user to mess up using the multiple helpers in an inconsistent way.

I guess we could implement that by adding a single event handler that captures key presses and prevents defaults only on certain key codes and target element ids. Would that make sense?


(Not asking you to follow this path but curious if it's something you considered and if it makes sense at all).

Reply:
Also... if we did this I guess we could move the close button configuration to an attribute instead of expecting the user to call `closeButton` function in the view and know that it will be placed using absolute positioning.

Reply:
🗒️ I wrote all of that before seeing the `FocusTrap` change. I think doing something among this lines might still make sense! Like.. the focus trap abstraction/module could be in charge of configuring this single event handler and that way it's even more decoupled from the view (assuming the ids are valid).

Reply:
I'm interested but not sure I'm following. When/how would we attach the tab and tab back listeners...?

Would you like to pair on this? Or if not, want to just do it? :) I like the API suggestions, as I understand them -- just not sure how to make them a reality

Reply:
I'll do a quick spike to see if it's actually doable and maybe ask you to pair on the API if it works!

Reply:
hmmm... it looks like it can't be done in pure Elm.

i was thinking about adding a single attribute to the top level container with something like:

```elm
multipleFocusableElements : { firstId : String, lastId: String, focus : String -> msg } -> Attribute msg
multipleFocusableElements {firstId, lastId, focus } =
  -- intention: use capture instead of bubble to run our event handler
  -- before the native focus change has occurred
  VirtualDom.onCapture "keydown"
  (VirtualDom.Custom
    (Decode.map3
      (\elementId keyCode shiftKey ->
          case (elementId, keyCode, shiftKey) of
            (firstId, tabKeyCode, True) ->
              { message : focus lastId
              , preventDefault : True
              , stopPropagation : True
              }
            (lastId, tabKeyCode, False) ->
              { message : focus firstId
              , preventDefault : True
              , stopPropagation : True
              }
            _ ->
              Decode.fail "..."
      )
      (Decode.at ["target", "id"] Decode.string)
      (Decode.field "key" Decode.int)
      (Decode.field "shiftKey" Decode.bool)
    )
  )
```

But the [VirtualDom docs](https://package.elm-lang.org/packages/elm/virtual-dom/latest/VirtualDom#on) say there is no support for capture phase events, and I don't think there's another way to prevent default tab behavior by attaching an event handler to a container element :-(